server/src/test/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderatorTester.java

```diff
@@ -0,0 +1,774 @@
+    private static final long DEFAULT_PUSH_TIMEOUT = 0;
+
+    private final AppendableIndexSpec appendableIndexSpec;
+    private final int maxRowsInMemory;
+    private final long maxBytesInMemory;
+    private final boolean skipBytesInMemoryOverheadCheck;
+    private final int maxColumnsToMerge;
+
+    // null if all partitionsSpec related params are null. see getDefaultPartitionsSpec() for details.
+    @Nullable
+    private final PartitionsSpec partitionsSpec;
+    private final IndexSpec indexSpec;
+    private final IndexSpec indexSpecForIntermediatePersists;
+    private final File basePersistDirectory;
+    private final int maxPendingPersists;
+
+    private final boolean forceGuaranteedRollup;
+    private final boolean reportParseExceptions;
+    private final long pushTimeout;
+    private final boolean logParseExceptions;
+    private final int maxParseExceptions;
+    private final int maxSavedParseExceptions;
+    private final long awaitSegmentAvailabilityTimeoutMillis;
+
+    @Nullable
+    private final SegmentWriteOutMediumFactory segmentWriteOutMediumFactory;
+
+    @Nullable
+    private static PartitionsSpec getPartitionsSpec(
+        boolean forceGuaranteedRollup,
+        @Nullable PartitionsSpec partitionsSpec,
+        @Nullable Integer maxRowsPerSegment,
+        @Nullable Long maxTotalRows,
+        @Nullable Integer numShards,
+        @Nullable List<String> partitionDimensions
+    )
+    {
+      if (partitionsSpec == null) {
+        if (forceGuaranteedRollup) {
+          if (maxRowsPerSegment != null
+              || numShards != null
+              || (partitionDimensions != null && !partitionDimensions.isEmpty())) {
+            return new HashedPartitionsSpec(maxRowsPerSegment, numShards, partitionDimensions);
+          } else {
+            return null;
+          }
+        } else {
+          if (maxRowsPerSegment != null || maxTotalRows != null) {
+            return new DynamicPartitionsSpec(maxRowsPerSegment, maxTotalRows);
+          } else {
+            return null;
+          }
+        }
+      } else {
+        if (forceGuaranteedRollup) {
+          if (!partitionsSpec.isForceGuaranteedRollupCompatibleType()) {
+            throw new IAE(partitionsSpec.getClass().getSimpleName() + " cannot be used for perfect rollup");
+          }
+        } else {
+          if (!(partitionsSpec instanceof DynamicPartitionsSpec)) {
+            throw new IAE("DynamicPartitionsSpec must be used for best-effort rollup");
+          }
+        }
+        return partitionsSpec;
+      }
+    }
+
+    public TestIndexTuningConfig(
+        Integer targetPartitionSize,
+        Integer maxRowsPerSegment,
+        AppendableIndexSpec appendableIndexSpec,
+        Integer maxRowsInMemory,
+        Long maxBytesInMemory,
+        Boolean skipBytesInMemoryOverheadCheck,
+        Long maxTotalRows,
+        Integer rowFlushBoundary_forBackCompatibility,
```

Comment:
Many parameters including this are deprecated in `IndexTuningConfig`. They only exist in `IndexTuningConfig` for compatibility. It doesn't seem reasonable to copy them to here.

Reply:
Clean config as much as I could, take a look.